spec/models/user_spec.rb

```diff
@@ -0,0 +1,77 @@
+require "rails_helper"
+
+RSpec.describe User, type: :model do
+
+  describe "associations" do
+    it {should have_many(:user_discounts).dependent(:destroy)}
+    it {should have_many(:ratings).dependent(:destroy)}
+    it {should have_many(:discounts).dependent(:destroy)}
+    it {should have_many(:payments).dependent(:destroy)}
+  end
+
+  describe "enum for status" do
+    it {should define_enum_for(:role).with_values(customer: 0, admin: 1)}
+  end
+
+  describe "validates" do
+    context "when validate presence" do
+      %i(email name phone_number password date_of_birth).each do |attr|
+        it {should validate_presence_of attr}
+      end
+    end
+
+    context "when length of name" do
+      it {should validate_length_of(:name).is_at_most(Settings.user.name.max_length)}
+    end
+    context "when length of password" do
+      it {should validate_length_of(:password).is_at_least(Settings.user.password.min_length)}
+    end
+    context "when length of phone" do
+      it {should validate_length_of(:phone_number).is_equal_to(Settings.user.phone_number.length)}
+    end
+
+    context "when validate unique of email" do
+      let!(:user_email1) {FactoryBot.create :user, email: "uSer1@gmail.com"}
+      let!(:user_email2) {FactoryBot.build :user, email: "uSer1@gmail.com"}
+      it {expect(user_email2.valid?).to eq false}
+    end
+
+    context "when email format" do
+      it {should allow_value("abcde@gmail.com").for(:email)}
+    end
+
+    context "when email wrong format" do
+      it {should_not allow_value("asdasdsa").for(:email)}
+    end
+  end
+
+  describe "downcase email" do
+    let!(:user_email) {FactoryBot.create :user, email: "USER@gmail.com"}
+    it "check before save downcase email" do
+      expect(user_email.email).to eq("user@gmail.com")
+    end
+  end
+
+  describe "Secure Password" do
+    it {should have_secure_password}
+  end
+
+  describe "#create_activation_digest" do
```

Comment:
describe, context dùng như thế nào e?

Reply:
describe thường dùng để nhóm các test case theo class, module, method 
context dùng để nhóm theo điều kiện của các test case
it là 1 test case